### PR TITLE
8087863: Mac: "Select All" within ListView/TreeView is handled differently depending on the useSystemMenuBar value

### PR DESCRIPTION
**Pull request by beldenfox**
macOS processes a shortcut key like Cmd+A in two phases. In the first phase it’s shopped around as a “key equivalent”. If it’s not consumed as a key equivalent it enters the second phase and processed as a normal keyDown event. Among other things the key equivalent phase ensures the shortcut will be seen by the system menu bar *before* being treated as a keyDown. This is the opposite of how JavaFX works; it expects a key event to be fired at the current focus node which gets first crack at the event before it works its way out to the menu bar.

We can’t really opt out of the key equivalent phase but we can get the event before the system menu bar does. Our implementation of performKeyEquivalent pushes the event through the JavaFX scene graph but has no way of knowing if the scene graph consumed it. The result is that a consumed event is always handed to the system menu bar where it can also trigger a menu item.

This PR introduces a variant of notifyKey that returns a boolean indicating whether the event was consumed or not. If the event was consumed performKeyEquivalent doesn’t allow it to continue on to the system menu bar.

I’m trying to fix this old, old problem because I’ve seen at least one JavaFX app tie itself up in knots trying to work around this. Despite the number of files being touched it is not a deep fix; there’s just a boolean return value that needs to be plumbed through multiple layers.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8087863](https://bugs.openjdk.org/browse/JDK-8087863): Mac: "Select All" within ListView/TreeView is handled differently depending on the useSystemMenuBar value (**Bug** - P4)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1528/head:pull/1528` \
`$ git checkout pull/1528`

Update a local copy of the PR: \
`$ git checkout pull/1528` \
`$ git pull https://git.openjdk.org/jfx.git pull/1528/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1528`

View PR using the GUI difftool: \
`$ git pr show -t 1528`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1528.diff">https://git.openjdk.org/jfx/pull/1528.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1528#issuecomment-2265987693)

### CODE REVIEW

**Comment by andy-goryachev-oracle** [comment]
Thinking out loud for a moment: wouldn't it be better to use a kind of EventFilter specifically for the shortcuts, before the normal dispatching happens?

**Comment by beldenfox** [comment]
> Thinking out loud for a moment: wouldn't it be better to use a kind of EventFilter specifically for the shortcuts, before the normal dispatching happens?

I think I see where you’re going here. We could send the KeyEvent into the scene graph and then catch it on the back end to be processed by the system menu bar. If the KeyEvent reaches the KeyboardAcceleratorHandler we know that it’s time for the menu bar to process it.

I’ve thought about this but it would tricky to implement. Currently we only send a subset of key events to the system menu bar (the one’s that are delivered to us via performKeyEquivalent). Once a KeyEvent is injected into the scene graph we lose track of whether it was generated by performKeyEquivalent or keyDown. We could also see KeyEvents that weren’t created by glass (anyone can create one). We would have to find a way to sort this out on the back end.

What truly makes me nervous is how we would have to alter the normal performKeyEquivalent flow. Usually NSApplication controls the order in which key equivalents are presented to different components (windows, views, the system menu bar, etc.) To implement this alternative approach we would have to short circuit the NSApplication flow and present events to the system menu bar manually.

This PR is the most conservative one I could come up with since it doesn’t change the order of event processing and fits into the normal performKeyEquivalent flow that NSApplication expects. It just avoids double-processing of the same event.

**Comment by beldenfox** [comment]
I'm marking this PR as draft for now. Turns out we're relying on double-processing of key events to make IM's work.

macOS sends a navigation key to us via performKeyEquivalent. That includes Home, End, and any arrow key. Currently we let the scene graph act on the event first and if a TextInput control has the focus the event is usually consumed. If we stop processing the event (and we should since it's consumed) it will never reach the IM.

In other words, the current system relies on having the same navigation key event sent to both performKeyEquivalent and keyDown. It gets acted on twice, once by the TextInput control and again by the IM. This is wrong but prioritizing the TextInput control over the IM is more wrong.

**Comment by andy-goryachev-oracle** [comment]
is there a compelling reason we are not using Event::isConsumed?

**Comment by beldenfox** [comment]
> is there a compelling reason we are not using Event::isConsumed?

As detailed in the discussion of PR #1523 the `isConsumed` flag doesn't work because the consumed event is a copy of the original one. The reliable way is to compare the result of the dispatch against null.

**Comment by andy-goryachev-oracle** [comment]
Maybe this case should also propagate `::isConsumed` flag similarly to #1523 

It looks to me we are trying to invent band-aids to what appears to be a design problem (cloning the events regardless of their `::isConsumed` state)

**Comment by beldenfox** [comment]
Taking this out of Draft. To ensure consistent handling of the IM both `keyDown:` and `performKeyEquivalent:` use the same routine. There's a check to ensure that the same event doesn't get processed twice which mimics an older check that used to be in the view delegate.

**Review by kevinrushforth — COMMENTED** [review]
The fix looks good to me. I'm finishing up my testing, but so far it is working as expected. Can you merge in the latest master?

**Review by kevinrushforth — APPROVED** [review]
Looks good.

**Comment by kevinrushforth** [comment]
@andy-goryachev-oracle Can you also review?

**Comment by andy-goryachev-oracle** [comment]
@kevinrushforth yes, it's in my queue

**Comment by andy-goryachev-oracle** [comment]
so here is an unrelated question - if we assume that the controls should only consume the event if it effected a change in said control, what should command-A do in case of a focused empty ListView?  should it consume the vent or not?

**Comment by beldenfox** [comment]
> so here is an unrelated question - if we assume that the controls should only consume the event if it effected a change in said control, what should command-A do in case of a focused empty ListView? should it consume the vent or not?

If the focus is clearly in the list view I think it would be surprising if Cmd+A was acted on further up the scene graph. Most of the discussion of event consumption has been around keys like ESCAPE where the user sometimes expects it to act locally (like canceling editing) and sometimes further out (like dismissing a dialog). I don't think that expectation applies to Cmd+A.

On the other hand a user might not perceive an empty list view as being the center of attention. It's not like an empty text field that has a blinking cursor.

File: modules/javafx.graphics/src/main/java/javafx/scene/Scene.java [hunk 1]
```diff
@@ -2192,9 +2193,11 @@ void processKeyEvent(KeyEvent e) {
         final EventTarget eventTarget =
                 (sceneFocusOwner != null && sceneFocusOwner.getScene() == Scene.this) ? sceneFocusOwner : Scene.this;
 
+        if (eventTarget == null) return false;
```
[the remark this answers is not in the document]
Comment (beldenfox) [reply modules/javafx.graphics/src/main/java/javafx/scene/Scene.java:2196]: This single-line pattern is used throughout this file. I was under the impression that it was within the coding guidelines but can't seem to find a citation for that.

File: modules/javafx.graphics/src/main/native-glass/mac/GlassView3D.m [hunk 2]
```diff
@@ -247,6 +247,8 @@ - (id)initWithFrame:(NSRect)frame withJview:(jobject)jView withJproperties:(jobj
         self->imEnabled = NO;
         self->handlingKeyEvent = NO;
         self->didCommitText = NO;
+
```
[the remark this answers is not in the document]
Comment (beldenfox) [reply modules/javafx.graphics/src/main/native-glass/mac/GlassView3D.m:250]: No, this is one-time initialization when the object is first created. We never re-initialize it so we should always set this field to nil.

File: tests/system/src/test/java/test/robot/javafx/scene/MenuDoubleShortcutTest.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+import com.sun.javafx.PlatformUtil;
+
+import test.util.Util;
+
+public class MenuDoubleShortcutTest {
+
+    static CountDownLatch startupLatch = new CountDownLatch(1);
+
+    static volatile Stage stage;
+    static volatile TestApp testApp;
```
[the remark this answers is not in the document]
Comment (beldenfox) [reply tests/system/src/test/java/test/robot/javafx/scene/MenuDoubleShortcutTest.java:58]: I am almost certainly copied that code from somewhere else so kudos to the original author :-)

File: modules/javafx.graphics/src/main/java/javafx/scene/Scene.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/scene/Scene.java:2196]: It is fine (and used all over the place).

OK:

```
    if (cond) return;

    if (cond) {
        return;
    }
```

NOT OK:

```
    if (cond)
        return;
```

**Review by andy-goryachev-oracle — APPROVED** [review]
thank you

**Review by kevinrushforth — APPROVED** [review]
